core/commands/storage.go

```diff
@@ -285,35 +292,44 @@ the chunk and replies back to client for the next challenge step.`,
 		// build connection with ledger
 		channelInfo, err := getChannelInfo(req.Context, channelID)
 		if err != nil {
-			ss.SetStatus(errState)
+			sm.Remove(ssID)
 			return err
 		}
 		log.Debug("Verified channel:", channelInfo)
 
 		go downloadChunkFromClient(chunkInfo, chunkHash, ssID, n, pid, req, env)
-		return res.Emit("Init success")
+		return nil
 	},
 }
 
+func removeSessionChunks(ssID string, chunkHash string) {
+	sm := storage.GlobalSession
+	ss := sm.Map[ssID]
```

Comment:
Unsafe, another goroutine might have removed it already .. cannot assume that the caller function is always the first one on this particular ssID/chunkHash

Can we make this whole part locked inside an `sm.RemoveChunks`?

Reply:
Why it is unsafe? I will check if ssID in map and chunkHash in session.chunk each time doing remove entry, and each remove process is requiring lock and release lock after that. If another goroutine has remove it. In remove step, nothing will happen, because I check if key is existed first

Reply:
Say two goroutines both enter `removeSessionChunks`, goroutine A executes until the end and removes it from `sm`. Then goroutine B executes `ss := sm.Map[ssID]` and `ss` is nil, and you call `ss.xxx` it will throw nil pointer.